core/src/network/handler/query.rs

```diff
@@ -5,31 +5,85 @@ use crate::{
-use std::{collections::BTreeSet, convert::TryInto, sync::Arc};
+use std::{convert::TryInto, sync::Arc};
 
 fn get_links(
     context: &Arc<Context>,
     base: Address,
     link_type: String,
     tag: String,
     crud_status: Option<CrudStatus>,
-) -> Vec<(Address, CrudStatus)> {
-    context
+    headers : bool
+) -> Result<Vec<GetLinkData>,HolochainError> {
+    //get links
+    let dht_store = context
         .state()
         .unwrap()
-        .dht()
+        .dht();
+    
+    let (get_link ,error) : (Vec<_>,Vec<_>) = dht_store
         .get_links(base, link_type, tag, crud_status)
-        .unwrap_or(BTreeSet::new())
+        .unwrap_or_default()
         .into_iter()
-        .map(|eav_crud| (eav_crud.0.value(), eav_crud.1))
-        .collect::<Vec<_>>()
+        .map(|eav_crud| (eav_crud.0.value(), eav_crud.1,eav_crud.0.attribute()))
+        //get tag
+        .map(|eav_crud|{
+            let tag = match eav_crud.2
+            {
+                Attribute::LinkTag(_,tag) => Ok(tag),
+                Attribute::RemovedLink(_,tag) => Ok(tag),
+                _ =>  Err(HolochainError::ErrorGeneric("Could not get tag".to_string()))
+            }.expect("INVALID ATTRIBUTE ON EAV GET, SOMETHING VERY WRONG IN EAV QUERY");
+            (eav_crud.0,eav_crud.1,tag)
+        })
+        //get targets from dht
+        .map(|eav_crud|{
+            dht_store
+            .get_link_targets(eav_crud.0.clone())
+            .map(|targets|{
+                targets.iter().last().map(|last_target|{
+                    Ok((eav_crud.0,eav_crud.1,last_target.value(),eav_crud.2))
+                }).unwrap_or(Err(HolochainError::ErrorGeneric("Could not find cached target".to_string())))
+            })
+            .unwrap_or(Err(HolochainError::ErrorGeneric("Could not find cached target".to_string())))
+        })
+        //get address from dht
+        .map(|address_crud_target|{
+            if headers
+            {
+                address_crud_target.map(|(address,crud_status,target,tag)|{
+                  dht_store
+                  .get_headers(address.clone())
+                  .map(|header|{
+                      Ok(GetLinkData::new(address,crud_status,target,tag,Some(header)))
+                  })
+                  .unwrap_or(Err(HolochainError::ErrorGeneric("Could not get headers".to_string())))
+              }).unwrap_or(Err(HolochainError::ErrorGeneric("Could not get headers".to_string())))
+            }
+            else
+            {
+                address_crud_target.map(|(address,crud_status,target,tag)|{GetLinkData::new(address,crud_status,target,tag,None)})
+            }
+        })
+        .partition(Result::is_ok);
+
+        //if can't find target throw error
+        if error.is_empty()
+        {
+            Ok(get_link.iter().map(|s|s.clone().unwrap()).collect::<Vec<_>>())
+        }
+        else
+        {
+            Err(HolochainError::ErrorGeneric("Could not find targets in local dht".to_string()))
```

Comment:
Might be nice to include the thrown errors in the output

Reply:
Good catch